vllm/v1/core/kv_cache_manager.py

```diff
@@ -36,41 +64,160 @@ def __init__(
+        #    same sequence), the one with more hash tokens (the tail of a block
+        #    chain) is at the front.
+        # We maintain this order by reversing the block order when free
+        # blocks of a request.
+        #
+        # Note that the block in this list is NOT guaranteed to be free
+        # due to prefix caching. If a block in free block list is touched
+        # by a request, we do not remove it immediately from free_block_list
+        # due to O(n) removal cost. Instead, we remove ref_cnt>0 blocks when
+        # 1. allocate new blocks in the same batch, or
+        # 2. finish scheduling a step and call async_remove_touched_blocks.
+        # That's why we need to maintain lazy_remove_block_ids and
+        # num_free_blocks counter separately.
+        #
+        # [No prefix caching] The free block list is simply in the order
+        # of last accessed time.
+        self.free_block_queue = deque(self.block_pool)
+        self.lazy_remove_block_ids = set()
+        self.num_free_blocks = num_gpu_blocks
+
+        self._async_executor = ThreadPoolExecutor(max_workers=1)
+        self._async_touch_task: Optional[Future] = None
+
+        # {block_hash: {block ID: block}}. A cached block is
+        # a full block with a block hash that can be used for prefix caching.
+        # The cached block may be used by running requests or in the
+        # free_block_queue that could potentially be evicted.
+        # NOTE: We currently don't de-duplicate the blocks in the cache,
+        # meaning that if a block becomes full and is cached, we don't check
+        # if there is already an identical block in the cache. This is because
+        # we want to make sure the allocated block IDs won't change so that
+        # block IDs are append-only.
+        self.cached_block_hash_to_block: Dict[int, Dict[
+            int, KVCacheBlock]] = defaultdict(dict)
+
+        # Mapping from request ID to block IDs to track the blocks allocated
+        # for each request, so that we can free the blocks when the request
+        # is finished.
         self.req_to_block_ids: Dict[str, List[int]] = {}
-        self.ref_cnts = np.zeros(num_gpu_blocks, dtype=np.int32)
 
     def get_computed_blocks(self, request: Request) -> List[int]:
+        """Get the computed (cached) blocks for the request.
+        Note that the computed blocks must be full.
+
+        Args:
+            request: The request to get the computed blocks.
+        
+        Returns:
+            A list of block IDs that are computed for the request.
+        """
         if not self.enable_caching:
-            # No prefix caching.
+            # Prefix caching is disabled.
             return []
-        # TODO(woosuk): Implement hash-based caching.
-        return []
+
+        computed_block_ids = []
+        block_hashes = self.hash_prompt_tokens(request.prompt_token_ids)
+
+        for block_hash in block_hashes:
```

Comment:
I think the invariance that if a block is in cache, all blocks before it must be in cache will allow a binary search here because a block's hash should also contain it's previous block.

Reply:
Yes this can be a good follow-up.